qf/group_test.go

```diff
@@ -19,3 +19,210 @@ func TestGetUserSubset(t *testing.T) {
 		t.Errorf("GetUserSubset() mismatch (-wantSubset, +gotSubset):\n%s", diff)
 	}
 }
+
+func TestGroup_UserNames(t *testing.T) {
+	t.Parallel()
+	tests := []struct {
+		name  string
+		group *qf.Group
+		want  []string
+	}{
+		{
+			name:  "Empty group",
+			group: &qf.Group{},
+			want:  nil,
+		},
+		{
+			name: "Non empty group",
+			group: &qf.Group{
+				Users: []*qf.User{
+					{Login: "adityaa"},
+					{Login: "tootsy-tiger"},
+					{Login: "rhea"},
+				},
+			},
+			want: []string{"adityaa", "tootsy-tiger", "rhea"},
+		},
+		{
+			name:  "Nil group",
+			group: nil,
+			want:  nil,
+		},
+		{
+			name: "Nil user",
+			group: &qf.Group{
+				Users: []*qf.User{nil},
+			},
+			want: nil,
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			got := tt.group.UserNames()
+			if diff := cmp.Diff(tt.want, got, protocmp.Transform()); diff != "" {
+				t.Errorf("Group.Usernames() mismatch (-wantSubset, +gotSubset):\n%s", diff)
+			}
+		})
+	}
+}
+
+func TestGroup_Contains(t *testing.T) {
+	t.Parallel()
+	tests := []struct {
+		name  string
+		group *qf.Group
+		user  *qf.User
+		want  bool
+	}{
+		{
+			name:  "Empty group",
+			group: &qf.Group{},
+			user:  &qf.User{ID: 1},
+			want:  false,
+		},
+		{
+			name: "User not in group",
+			group: &qf.Group{
+				Users: []*qf.User{
+					{ID: 1},
+					{ID: 2},
+				},
+			},
+			user: &qf.User{ID: 3},
+			want: false,
+		},
+		{
+			name: "User in group",
+			group: &qf.Group{
+				Users: []*qf.User{
+					{ID: 1},
+					{ID: 2},
+				},
+			},
+			user: &qf.User{ID: 2},
+			want: true,
+		},
+		{
+			name: "Nil user",
+			group: &qf.Group{
+				Users: []*qf.User{
+					{ID: 1},
+					{ID: 2},
+					{ID: 3},
+				},
+			},
+			user: nil,
+			want: false,
+		},
+		{
+			name:  "Nil group and nil user",
+			group: nil,
+			user:  nil,
+			want:  false,
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			got := tt.group.Contains(tt.user)
+			if diff := cmp.Diff(tt.want, got, protocmp.Transform()); diff != "" {
+				t.Errorf("Group.Contains() mismatch (-wantSubset, +gotSubset):\n%s", diff)
+			}
+		})
+	}
+}
+
+func TestGroup_ContainsAll(t *testing.T) {
+	t.Parallel()
+	tests := []struct {
+		name string
+		g1   *qf.Group
+		g2   *qf.Group
+		want bool
+	}{
+		{
+			name: "Two empty groups",
+			g1:   &qf.Group{},
+			g2:   &qf.Group{},
+			want: true,
+		},
+		{
+			name: "Identical groups",
+			g1: &qf.Group{
+				Users: []*qf.User{
+					{ID: 1, Name: "user-1"},
+					{ID: 2, Name: "user-2"},
+				},
+			},
+			g2: &qf.Group{
+				Users: []*qf.User{
+					{ID: 1, Name: "user-1"},
+					{ID: 2, Name: "user-2"},
+				},
+			},
+			want: true,
+		},
+		{
+			name: "Different order same users",
```

Comment:
This test is failing because of `reflect.DeepEqual()`. As this also compares if the order is same. I am not sure if we should handle this case or not?

Reply:
I am quite new to the project so I might need help figuring some things out.

Reply:
@meling @JosteinLindhom can you please suggest?